### PR TITLE
kubeadm: fixed etcd sync endpoints

### LINKED ISSUE #1300: kubeadm reset success but this node ip still in kubeadm-config configmap
## Is this a BUG REPORT or FEATURE REQUEST?

BUG REPORT

## Versions

**kubeadm version** (use `kubeadm version`):
```sh
[root@k8s-211 ~]# kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.0", GitCommit:"ddf47ac13c1a9483ea035a79cd7c10005ff21a6d", GitTreeState:"clean", BuildDate:"2018-12-03T21:02:01Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
```
**Environment**:
- **Kubernetes version** (use `kubectl version`):
```sh
[root@k8s-211 ~]# kubectl version
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.0", GitCommit:"ddf47ac13c1a9483ea035a79cd7c10005ff21a6d", GitTreeState:"clean", BuildDate:"2018-12-03T21:04:45Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.0", GitCommit:"ddf47ac13c1a9483ea035a79cd7c10005ff21a6d", GitTreeState:"clean", BuildDate:"2018-12-03T20:56:12Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
```
- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release):
```sh
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```
- **Kernel** (e.g. `uname -a`):
```sh
Linux k8s-lixin-211 3.10.0-693.el7.x86_64 #1 SMP Tue Aug 22 21:09:27 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```
- **Others**:

## What happened?

I use `kubeadm reset -f` to reset this control plane node, the command run success. But when i see `kubeadm-config` ConfigMap, it already have this node ip in ClusterStatus.

I still have a question, why `kubeadm reset` not delete this node directly from the cluster? Instead, run `kubectl delete node <node name>` manually.

## What you expected to happen?

`kubeadm-config` ConfigMap remove this node ip.

## How to reproduce it (as minimally and precisely as possible)?

- `kubeadm init --config=kubeadm.yml` on the first node.
- `kubeadm join --experimental-control-plane --config=kubeadm.yml` on the second node.
- `kubeadm reset -f` on the second node.
- `kubectl -n kube-system get cm kubeadm-config -oyaml` find the second node ip already in ClusterStatus.

## Anything else we need to know?

<details>
<summary>kubeadm-config configMap yaml:</summary>

```yaml
apiVersion: v1
data:
  ClusterConfiguration: |
    apiServer:
      extraArgs:
        authorization-mode: Node,RBAC
      timeoutForControlPlane: 4m0s
    apiVersion: kubeadm.k8s.io/v1beta1
    certificatesDir: /etc/kubernetes/pki
    clusterName: kubernetes
    controlPlaneEndpoint: 192.168.46.117:6443
    controllerManager: {}
    dns:
      type: CoreDNS
    etcd:
      local:
        dataDir: /var/lib/etcd
        extraArgs:
          cipher-suites: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
        serverCertSANs:
        - 192.168.46.117
    imageRepository: k8s.gcr.io
    kind: ClusterConfiguration
    kubernetesVersion: v1.13.0
    networking:
      dnsDomain: cluster.local
      podSubnet: 10.244.0.0/16
      serviceSubnet: 10.96.0.0/12
    scheduler: {}
  ClusterStatus: |
    apiEndpoints:
      k8s-211:
        advertiseAddress: 192.168.46.211
        bindPort: 6443
      k8s-212:
        advertiseAddress: 192.168.46.212
        bindPort: 6443
    apiVersion: kubeadm.k8s.io/v1beta1
    kind: ClusterStatus
kind: ConfigMap
metadata:
  creationTimestamp: "2018-12-04T14:17:38Z"
  name: kubeadm-config
  namespace: kube-system
  resourceVersion: "103402"
  selfLink: /api/v1/namespaces/kube-system/configmaps/kubeadm-config
  uid: 5a9320c1-f7cf-11e8-868d-0050568863b3
```
</details>

### PR DESCRIPTION
**Pull request by pytimer**
**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
> /kind bug

**What this PR does / why we need it**:

This PR fix kubeadm sync etcd endpoints problem.

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubeadm/issues/1300

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: fix a bug when syncing etcd endpoints
```

### CODE REVIEW

**Review by rosti — APPROVED** [review]
Well spotted @pytimer ! Thank you!
Can you also create cherry-pick PR for 1.13 once this one merges?

/lgtm
/ok-to-test
/kind bug
/assign @fabriziopandini 
/assign @neolit123

File: cmd/kubeadm/app/util/etcd/etcd.go
Comment (neolit123) [inline cmd/kubeadm/app/util/etcd/etcd.go]: having a single method that accepts a pointer and the rest accepting non-pointers seems a bit odd.
we might want to make all of them accept pointer... see GetVersion(),  GetClusterVersions() etc.

**Review by neolit123 — COMMENTED** [review]
@pytimer 
thanks, added a comment about pointers vs non-pointers.

File: cmd/kubeadm/app/util/etcd/etcd.go
Comment (pytimer) [reply cmd/kubeadm/app/util/etcd/etcd.go]: you mean that Sync() return the real etcd endpoints? such as:

```go
func (c Client) Sync() ([]string, error) {}
```

Comment (neolit123) [reply cmd/kubeadm/app/util/etcd/etcd.go]: the methods for `Client` have different signatures:
`func (c *Client) Sync()`
`func (c Client) AddMember(name string, peerAddrs string) ([]Member, error) {`
...

`Sync()` uses a pointer, but the rest of the methods use non-pointers.

Comment (pytimer) [reply cmd/kubeadm/app/util/etcd/etcd.go]: Sync() only return error now, but the rest of the methods return multi values.
Can change this `Sync()` method return endpoints and error, like other methods?
The `NewFromCluster` invoke `Sync()` like below:

```go
endpoints, err = etcdClient.Sync()
if err != nil {
	return nil, errors.Wrap(err, "error syncing endpoints with etc")
}
etcdClient.Endpoints = endpoints
```
In this case, all methods use non-pointers, and only change `Sync()` methods, but test case maybe need to update.

The other ways, the methods of `Client` all change non-pointers to pointer, the way need to change more codes.

I am not sure if i understand what you mean？If wrong, can you explain your comments again. Thanks.

Comment (neolit123) [reply cmd/kubeadm/app/util/etcd/etcd.go]: @pytimer my point is this:
```
func (c *Client) Sync(...) ...
func (c Client) AddMember(...) ...
...
```
^ see how `c` is a pointer in `Sync` but not a pointer in the rest of methods for the `Client` type.
we need to make all the methods consistent.

Comment (fabriziopandini) [reply cmd/kubeadm/app/util/etcd/etcd.go]: @pytimer I'm ok with the solution you are proposing:
```
func (c Client) Sync() ([]string, error) {}
```

@neolit123 IMO the solution above keeps consistency in the signature and address the issue.
Eventually we will reconsider changing all the methods in another PR

Comment (neolit123) [reply cmd/kubeadm/app/util/etcd/etcd.go]: /approve
as per @fabriziopandini 's comment.
thanks @pytimer

Comment (fabriziopandini) [reply cmd/kubeadm/app/util/etcd/etcd.go]: /hold

Comment (fabriziopandini) [reply cmd/kubeadm/app/util/etcd/etcd.go]: @pytimer please fix as discussed above

Comment (pytimer) [reply cmd/kubeadm/app/util/etcd/etcd.go]: @fabriziopandini @neolit123 

Fix like:

```go
func (c Client) Sync() ([]string, error) {}
```

Is it ok?

**Comment by pytimer** [comment]
@neolit123 @fabriziopandini I send a new commit https://github.com/kubernetes/kubernetes/pull/71945/commits/ca1fdbc11565a03525aaec7249f5f25822f62a84 ， do i still need to other changes?

If need change, hope point out. Thanks.

File: cmd/kubeadm/app/util/etcd/etcd.go
Comment (rosti) [reply cmd/kubeadm/app/util/etcd/etcd.go]: @pytimer @fabriziopandini @neolit123 I am not OK with `func (c Client) Sync() ([]string, error)`. This is no longer a `Sync` method. It's not even a `GetEndpoints` method (because we have `Endpoints` member in `Client`). Furthermore, this introduces a code smell (inappropriate intimacy), where `NewFromCluster` now updates field inside `Client`.

This is a lot worse solution to the problem it tries to solve. I am much more in favor of making all methods `func (c *Client)`, than have this. I am also in favor of having only `Sync()` take a pointer to `Client`.

**Comment by pytimer** [comment]
@fabriziopandini Like @rosti said, he think `func (c Client) Sync() ([]string, error)` maybe not good, so whether needs other changes?

**Comment by fabriziopandini** [comment]
@pytimer thanks for pointing me to @rosti latest comment!

I kindly ask you to have a little be more of patience while we agree on what is the best solution that allows a small and focused fix that can be safely cherry-picked and also preserve code quality.
As soon as @neolit123 shows up later today I will try to move this discussion on zoom/slack to get this moving asap.

In the meantime, thanks again for your availability and responsiveness. Really appreciated!

File: cmd/kubeadm/app/util/etcd/etcd.go
Comment (neolit123) [reply cmd/kubeadm/app/util/etcd/etcd.go]: @pytimer 
OK, i missed the part that we want to modify the existing Client object and not a copy.
please bring back `(c *Client)` .

**Comment by neolit123** [comment]
@pytimer 
added a comment here: https://github.com/kubernetes/kubernetes/pull/71945#discussion_r241005342

please also make the release note in the OP:
`kubeadm: fix a bug when syncing etcd endpoints`
instead of `NONE`.

thanks.

**Comment by pytimer** [comment]
@neolit123 @fabriziopandini @rosti 

I bring back `(c *Client) Sync()`, and edit comment according to @neolit123 feedback. Can you check it out?

Please let me know if i need to do other things. thanks.

**Comment by neolit123** [comment]
@pytimer 
thanks for the update.

**Comment by fabriziopandini** [comment]
@pytimer thanks!
/hold cancel

**Comment by rosti** [comment]
@pytimer looks superb again! Thank you for this!
We have to do a cherry-pick of this change for 1.13. If you can't handle this yourself, I can do it. Just let me know.

/lgtm

**Comment by pytimer** [comment]
@rosti I have no experience in using cherry-pick on the kubernetes repository, so can you cherry-pick this change? Thanks.

Can you tell me how to cherry-pick on the kubernetes repository if you have time? I want to study it and if there is a similar operation in the future, i can do it. Thanks.

**Comment by rosti** [comment]
@pytimer please, follow this guide. It's really simple. If you have issues you can ping me to do it. Thanks!
https://github.com/kubernetes/community/blob/master/contributors/devel/cherry-picks.md

**Comment by pytimer** [comment]
@rosti Thanks for you help about cherry-pick. I already cherry-pick this PR. 

Cherry pick is https://github.com/kubernetes/kubernetes/pull/72042.

**Pull request merged**
